src/keychain/index.ts

```diff
@@ -0,0 +1,271 @@
+import DeviceInfo from 'react-native-device-info';
+import {
+  ACCESS_CONTROL,
+  ACCESSIBLE,
+  AUTHENTICATION_TYPE,
+  canImplyAuthentication,
+  getAllInternetCredentials,
+  getInternetCredentials,
+  getSupportedBiometryType as originalGetSupportedBiometryType,
+  hasInternetCredentials,
+  Options,
+  resetInternetCredentials,
+  setInternetCredentials,
+  UserCredentials,
+  BIOMETRY_TYPE,
+  requestSharedWebCredentials,
+  setSharedWebCredentials as originalSetSharedWebCredentials,
+  SharedWebCredentials,
+} from 'react-native-keychain';
+import { MMKV } from 'react-native-mmkv';
+
+import { delay } from '@/helpers/utilities';
+import { IS_DEV, IS_IOS } from '@/env';
+import { logger } from '@/logger';
+
+export enum ErrorType {
+  Unknown = 0,
+  UserCanceled = -1, // legacy
+  NotAuthenticated = -2, // legacy
+  Unavailable = -3,
+}
+
+type Result<T = any> =
+  | {
+      value: T;
+      error: undefined;
+    }
+  | {
+      value: undefined;
+      error: ErrorType;
+    };
+
+const cache = new MMKV({
+  id: 'rainbowKeychainLocalStorage',
+});
+
+export const publicAccessControlOptions: Options = {
+  accessible: ACCESSIBLE.WHEN_UNLOCKED_THIS_DEVICE_ONLY,
+};
+
+export async function get(
+  key: string,
+  options: Options = {},
+  attempts = 0
+): Promise<Result<string>> {
+  logger.debug(`keychain: get`, { key }, logger.DebugContext.keychain);
+
+  let data = cache.getString(key);
+
+  if (!data) {
+    try {
+      const result = await getInternetCredentials(key, options);
+
+      if (result) {
+        data = result.password;
+      }
+    } catch (e: any) {
+      switch (e.toString()) {
+        case 'Error: User canceled the operation.': {
+          return {
+            value: undefined,
+            error: ErrorType.UserCanceled,
+          };
+        }
+        case 'Error: Wrapped error: User not authenticated': {
+          return {
+            value: undefined,
+            error: ErrorType.NotAuthenticated,
+          };
+        }
+        case 'Error: The user name or passphrase you entered is not correct.': {
+          if (attempts > 2) {
+            return {
+              value: undefined,
+              error: ErrorType.NotAuthenticated,
+            };
+          }
+
+          // try again
+          await delay(1000);
+          return get(key, options, attempts + 1);
+        }
+        default: {
+          return {
+            value: undefined,
+            error: ErrorType.Unknown,
+          };
+        }
+      }
+    }
+  }
+
+  return data
+    ? {
+        value: data,
+        error: undefined,
+      }
+    : {
+        value: undefined,
+        error: ErrorType.Unavailable,
+      };
+}
+
+export async function set(
+  key: string,
+  value: string,
+  options: Options = {}
+): Promise<void> {
+  logger.debug(`keychain: set`, { key }, logger.DebugContext.keychain);
+
+  // only save public data to mmkv
```

Comment:
maybe it's worth mentioning somewhere in the readme that the keychain wrapper is currently a Frankenstein between native keychain and mmkv  and why.

CC: @skylarbarrera

Reply:
Good call 👍